### PR TITLE
[Merged by Bors] - fix: mdata in abel

### PR DESCRIPTION
**Pull request by adomani**
PR inspired by "automated bugs" (#12054).

[Zulip thread](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/automated.20bugs/near/432861117)

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by semorrison** [comment]
bors d+

File: Mathlib/Tactic/Abel.lean [hunk 1]
```diff
@@ -473,7 +473,7 @@ open Elab.Tactic Parser.Tactic
 /-- Use `abel_nf` to rewrite the main goal. -/
 def abelNFTarget (s : IO.Ref AtomM.State) (cfg : AbelNF.Config) : TacticM Unit := withMainContext do
   let goal ← getMainGoal
-  let tgt ← instantiateMVars (← goal.getType)
+  let tgt ← instantiateMVars (← goal.getType')
```
Comment (kmill) [inline Mathlib/Tactic/Abel.lean:476]: I think this is too strong (it does whnf at default transparency).

I'm not sure about doing reduction (I know @fpvandoorn is not sold on doing whnf before matching). There's also `cleanupAnnotations (← instatiateMVars (← goal.getType))`.

Comment (fpvandoorn) [reply Mathlib/Tactic/Abel.lean:476]: In the following code snippet, I would indeed hope that `abel_nf` doesn't unfold the definition of `MyTrue`.
```
def MyTrue := True
example : MyTrue := by
  abel_nf
```

Comment (eric-wieser) [reply Mathlib/Tactic/Abel.lean:476]: It should unfold it if it is `@[reducible]` though, right?

Comment (adomani) [reply Mathlib/Tactic/Abel.lean:476]: Would `← whnfR (← instantiateMVars (← goal.getType'))` and then guard-messaging the test suggested by Floris as an error be a solution with which everyone is happy?

Comment (kmill) [reply Mathlib/Tactic/Abel.lean:476]: Watch out, whnfR has to occur before instantiateMVars.

Better I think would be withReducible around the current code. (With getType')

Comment (adomani) [reply Mathlib/Tactic/Abel.lean:476]: Kyle, assuming that
```lean
  let tgt ← Lean.Meta.withReducible (instantiateMVars (← goal.getType'))
```
is the change that you suggested, then Floris' example actually gets solved by `abel_nf`.

Comment (kmill) [reply Mathlib/Tactic/Abel.lean:476]: Sorry, I made a mistake with what I wrote, I wrote it on my phone. You need to be careful with `withReducible` and `←` lifting. The `getType'` function doesn't need `instantiateMVars`, so with that simplification, the correct version should be
```lean
  let tgt ← withReducible goal.getType'
```

Comment (adomani) [reply Mathlib/Tactic/Abel.lean]: Thanks Kyle: I used your latest version and it passes the tests, both successfully and not!

**Pull request closed without merging**